rllib/core/rl_trainer/trainer_runner.py

```diff
@@ -0,0 +1,181 @@
+        self._compute_config = compute_config
+
+        # TODO: remove the _compute_necessary_resources and just use
+        # trainer_config["use_gpu"] and trainer_config["num_workers"]
+        resources = self._compute_necessary_resources()
+        scaling_config = ScalingConfig(
+            num_workers=resources["num_workers"],
+            use_gpu=resources["use_gpu"],
+        )
+        # the only part of this class that is framework agnostic:
+        if framework == "torch":
+            from ray.train.torch import TorchConfig
+
+            backend_config = TorchConfig()
+        elif framework == "tf":
+            from ray.train.tensorflow import TensorflowConfig
+
+            backend_config = TensorflowConfig()
+        else:
+            raise ValueError("framework must be either torch or tf")
+
+        self.backend_executor = BackendExecutor(
+            backend_config=backend_config,
+            num_workers=scaling_config.num_workers,
+            num_cpus_per_worker=scaling_config.num_cpus_per_worker,
+            num_gpus_per_worker=scaling_config.num_gpus_per_worker,
+            max_retries=0,  # TODO: make this configurable in trainer_config
+            # with default 0
+        )
+
+        # TODO: let's not pass this into the config which will cause
+        # information leakage into the SARLTrainer about other workers.
+        scaling_config = {"world_size": resources["num_workers"]}
+        trainer_config["scaling_config"] = scaling_config
+        trainer_config["distributed"] = bool(self._compute_config["num_gpus"] > 1)
+        self.backend_executor.start(
+            train_cls=trainer_class, train_cls_kwargs=trainer_config
+        )
+        self.workers = [w.actor for w in self.backend_executor.worker_group.workers]
+
+        ray.get([w.init_trainer.remote() for w in self.workers])
+
+    def _compute_necessary_resources(self):
+        num_gpus = self._compute_config.get("num_gpus", 0)
+        num_workers = self._compute_config.get("num_training_workers", 0)
+        if num_workers and num_gpus:
+            assert num_workers == num_gpus, (
+                "If num_training_workers and "
+                "num_gpus are specified it must be equal to num_gpus"
+            )
+
+        elif num_gpus and not num_workers:
+            num_workers = num_gpus
+
+        elif not num_gpus and not num_workers:
+            num_workers = 1
+
+        return {"num_workers": num_workers, "use_gpu": bool(num_gpus)}
+
+    def update(self, batch: MultiAgentBatch = None, **kwargs):
+        """TODO: account for **kwargs
+        Example in DQN:
+            >>> trainer_runner.update(batch) # updates the gradient
+            >>> trainer_runner.update(update_target=True) # should soft-update
+                the target network
+        """
+        refs = []
+        if batch is None:
+            for worker in self.workers:
+                refs.append(worker.update.remote(**kwargs))
```

Comment:
can you use FaultToleranceActorManager to manage these workers please?
and do ``foreach_worker()`` here instead?
otherwise, you need to catch actor errors yourself.
maybe a TODO also works for now, but this may be a pretty important TODO in practice.

Reply:
So I added todos in the init function. Ive just added this class for now mainly as a place holder until we come back and do everything we need to for this class.